### DISCUSSION TITLE
What's the best way to disable a next/link if the href matches the current URL?

### DISCUSSION THREAD

**Discussion by transitive-bullshit** [vercel/next.js#42706] [Help] [answered]
Using `next/link`, what's the best practice for not including an `<a>` link (or just disabling it) if its target `href` matches the current route?

I'm wondering if there's an official solution, since it seems like a very common pattern, or if we should just duplicate the JSX for the linked and non-linked versions?

Thanks!

**Comment by icyJoseph** [comment]
> Using next/link, what's the best practice for not including an <a> link (or just disabling it) if its target href matches the current route?

What do you mean with this? What kind of HTML elements do you expect to see when the href matches the current route. Is this some kind of reverse active link behavior?

There's this example from the docs on how to detect active links: https://github.com/vercel/next.js/blob/canary/examples/active-class-name/components/ActiveLink.tsx I think you could tweak it to do what you want.

Do remember that `a` tags have no disable attribute, unlike buttons, so you gotta apply CSS to disable pointer events (`none`). And users can still tab into it... it is often best to also intercept the event and prevent it's default behaviour.

```jsx
// pseudocode
let disabled = true // you have to figure out how to calculate this, perhaps the active link helps out

// later on somewhere in the JSX
          <Link href="/some-path" legacyBehavior>
            <a
              style={disabled ? { pointerEvents: 'none' } : {}}
              onClick={(e) => {
                if (disabled) {
                  e.preventDefault();
                }
              }}
            >
              some-path
            </a>
          </Link>
```  

> The disabled attribute is supported by [button](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button), command, [fieldset](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/fieldset), [keygen](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/keygen), [optgroup](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/optgroup), [option](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/option), [select](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select), [textarea](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/textarea) and [input](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input).

**Reply by transitive-bullshit** [reply]
Thanks @icyJoseph, this is definitely along the lines of what I'm looking for.

I'm using Next.js 13, so I'll need to tweak the `ActiveLink` example a bit, but I'm fine with just disabling the `<a>` instead of replacing it with a `<span>` or something.

Will post my solution when I get it working. Thanks again!

**Comment by transitive-bullshit** [comment, marked as answer]
This is the solution I came up with for Next.js 13, inspired by @icyJoseph's answer.

The name `ActiveLink` probably isn't the most appropriate; it will disable the link if it's active.

```ts
'use client'

import * as React from 'react'
import { usePathname } from 'next/navigation'
import Link, { LinkProps } from 'next/link'
import cs from 'clsx'

type ActiveLinkProps = LinkProps & {
  children: React.ReactNode
  className?: string
  activeClassName?: string
  style?: React.CSSProperties
}

/**
 * Link that will be disabled if the target `href` is the same as the current
 * route's pathname.
 */
export const ActiveLink = ({
  children,
  href,
  style,
  className,
  activeClassName,
  onClick,
  ...props
}: ActiveLinkProps) => {
  const pathname = usePathname()
  const [disabled, setDisabled] = React.useState(false)

  React.useEffect(() => {
    const linkPathname = new URL(href as string, location.href).pathname

    setDisabled(linkPathname === pathname)
  }, [pathname, href])

  const styleOverride = React.useMemo<React.CSSProperties>(
    () =>
      disabled
        ? {
            pointerEvents: 'none',
            ...style
          }
        : style ?? {},
    [disabled, style]
  )

  const onClickOverride = React.useCallback(
    (event: React.MouseEvent) => {
      if (disabled) {
        event.preventDefault()
        return false
      }

      if (onClick) {
        return onClick(event)
      }
    },
    [disabled, onClick]
  )

  return (
    <Link
      {...props}
      className={cs(className, disabled && activeClassName)}
      href={href}
      style={styleOverride}
      onClick={onClickOverride}
    >
      {children}
    </Link>
  )
}
```
